### PR TITLE
segresnet algo update

### PR DESCRIPTION
**Pull request by myron**
Intermediate update of segresnet algo,  until it's merged in the core.
compatible with all existing workflows.

### CODE REVIEW

File: auto3dseg/algorithm_templates/segresnet/configs/hyper_parameters.yaml [hunk 1]
```diff
@@ -1,63 +1,78 @@
...
-  optimizer: "$@optimizer"
-  warmup_epochs: 20
-  warmup_multiplier: 0.1
-  T_max: "@num_epochs"
-
-# fine-tuning
+  weight_decay: 1.e-5
+
+network:
+  _target_: SegResNetDS
```
Comment (KumoLiu) [inline auto3dseg/algorithm_templates/segresnet/configs/hyper_parameters.yaml:54]: Hi @myron, I noticed that you have changed the loss function, removed lr_scheduler, and used a new version SegResNet. Not sure if it makes more sense to change the name of this whole template to segresnetds or just add a new template to keep the original segresnet if there are users who still want to use the original version. What do you think? 
In addition to that, perhaps we need to update the results in the tutorial repo. 
https://github.com/Project-MONAI/tutorials/tree/main/auto3dseg/tasks/msd
Thanks!

Comment (myron) [reply auto3dseg/algorithm_templates/segresnet/configs/hyper_parameters.yaml:54]: thank you for the comment. the loss function is the same (it's just a loss with deep supervision), and the network is the same (but with a deep supervision support).  I don't think we need another template, it'll be difficult to maintain.  and we provide a versioned tar.gz release url to download the algo templates (if anyone wants to use the older variant)

regarding updates to the tutorials repo,  we can can consider that after this is merged. thank you

Comment (KumoLiu) [reply auto3dseg/algorithm_templates/segresnet/configs/hyper_parameters.yaml:54]: Hi @myron, thanks for your explanation, I know the network is the same then maybe we could change this `dsdepth` to 1? Users may not be able to notice that this new version of the template has added deep supervision support.
https://github.com/Project-MONAI/research-contributions/blob/fde97625e100e1d44d029fcc98186efd1a2a190c/auto3dseg/algorithm_templates/segresnet/configs/hyper_parameters.yaml#L60

And why do you need to remove the learning rate scheduler？ For better performance or anything else? Could you please help explain it? 
I noticed that you have simplified the config, putting all the configs in the same file. This does look clearer and simpler. I remember we had a similar discussion before about why we need to separate the config into different files, it seems to prevent overwriting problems caused by duplicate parameters. Maybe there is no such problem in this template, I'm not sure if we could keep both ways of writing the config. Could you please help confirm it? @dongyang0122 @mingxin-zheng @wyli, Thanks!

Comment (myron) [reply auto3dseg/algorithm_templates/segresnet/configs/hyper_parameters.yaml:54]: if you notice that some options such as dsdepth have changed, it's because they are updated to be the proper  values, that we should have used from the beginning, and here we're fixing them. 

regarding the learning rate scheduler. this particular "algorithm template" is a full workflow , it's not just a network/model with arbitrary training options. and in this workflow we use a particular learning rate scheduler.  Same for many other components used during training. We find a balance, by exposing a large set of parameters, but not every single class. 

if the config file is very long, it is split into several files purely for visual convenience (they are all merged during loading). Here, there is no need to split into sub configs. thanks.

Comment (mingxin-zheng) [reply auto3dseg/algorithm_templates/segresnet/configs/hyper_parameters.yaml:54]: I saw a style change in the way to organize the code. For example, transforms are taken out from the bundle configs. I think this is the new direction after discussions and I am completely fine with it. Thanks!

File: auto3dseg/algorithm_templates/segresnet/scripts/algo.py [hunk 2]
```diff
@@ -34,292 +65,194 @@ def fill_template_config(self, data_stats_file, output_path, **kwargs):
...
 
+    def export_to_disk(self, output_path: str, algo_name: str, **kwargs):
+        super().export_to_disk(output_path=output_path, algo_name=algo_name, **kwargs)
 
-if __name__ == "__main__":
-    from monai.utils import optional_import
+        output_path =os.path.join(output_path, algo_name)
+        config = ConfigParser.load_config_file(os.path.join(output_path, "configs/hyper_parameters.yaml"))
+
+        for c in config.get('custom_data_transforms',[]):
```
Comment (mingxin-zheng) [inline auto3dseg/algorithm_templates/segresnet/scripts/algo.py:248]: The `custom_data_transforms` will be saved in the the `configs/hyper_parameters.yaml` of the generated algorithm folder (e.g. `segresent_0`). It means it will be used when users do the command-line-based trainings
`python segresnet_0/train.py` or `torchrun segresnet_0/train.py`

But I am worried that the following way will not load the `custom_data_transforms`
```
one_algo = algo_from_pickle('segresnet.pkl')
one_algo.train()
```
Because the process uses the templates and the `fill_records` for instantiation.

File: auto3dseg/algorithm_templates/segresnet/scripts/validate.py [hunk 3]
```diff
@@ -9,204 +9,20 @@
...
-        dict_file = {}
-        dict_file["acc"] = float(avg_metric)
-        for _c in range(metric_dim):
-            dict_file["acc_class" + str(_c + 1)] = metric[2 * _c] / metric[2 * _c + 1]
-
-        with open(os.path.join(output_path, "summary.yaml"), "w") as out_file:
-            yaml.dump(dict_file, stream=out_file)
-
-    return
+    override["validate#enabled"] = True
```
Comment (mingxin-zheng) [inline auto3dseg/algorithm_templates/segresnet/scripts/validate.py:23]: If the users set `validate#enabled` to False with `override` but we need to override their `override`, can we make an info-level or debug-level log?

File: auto3dseg/algorithm_templates/segresnet/scripts/algo.py (hunk 2 above)
Comment (myron) [reply auto3dseg/algorithm_templates/segresnet/scripts/algo.py:248]: thanks, I'm not exactly sure what you mean, it seems to work fine for me.  if you found it not working, please suggest a change.   But overall, we do have a tentative plan to remove pickling strategy in the future.

File: auto3dseg/algorithm_templates/segresnet/scripts/validate.py (hunk 3 above)
Comment (myron) [reply auto3dseg/algorithm_templates/segresnet/scripts/validate.py:23]: the default is always False, if we override, it becomes a new rule (it's the same for any other parameter). why do we need to log anything.   PS: may be it's not clear what it means, when it's true it will run 1 validation epoch over the validation set (no training or anything else will be performed)

**Comment by mingxin-zheng** [comment]
I saw some items are printed twice when I run the new segresnet in my local environment. @myron Can you check if the outputs are all necessary? Thanks!
[segresnet.txt](https://github.com/Project-MONAI/research-contributions/files/10163271/segresnet.txt)

File: auto3dseg/algorithm_templates/segresnet/scripts/algo.py (hunk 2 above)
Comment (mingxin-zheng) [reply auto3dseg/algorithm_templates/segresnet/scripts/algo.py:248]: I'm not sure how "custom_data_transforms" will look like. Something below could be a fix in the end of the reloaded `export_to_disk` functions.

```python
# to ensure the exported pickle file will include information about the custom_data_transform
self.fill_records['hyper_parameters.yaml']['custom_data_transforms'] = ...
```

File: auto3dseg/algorithm_templates/segresnet/scripts/validate.py (hunk 3 above)
Comment (mingxin-zheng) [reply auto3dseg/algorithm_templates/segresnet/scripts/validate.py:23]: Okay I see what you are saying. I assume the users won't likely do something like:
```
python validate.py --validate#enabled=False
```

File: auto3dseg/algorithm_templates/segresnet/scripts/segmenter.py [hunk 4]
```diff
@@ -0,0 +1,1395 @@
...
+
+
+def run_segmenter(config_file: Optional[Union[str, Sequence[str]]] = None, **kwargs):
+    """
+    if multiple gpu available, start multiprocessing for all gpus
+    """
+    nprocs = torch.cuda.device_count()
+    if nprocs > 1 and not dist.is_torchelastic_launched():
+        kwargs["mgpu"] = {"world_size": nprocs, "init_method": kwargs.get("init_method", "tcp://127.0.0.1:23456")}
+        torch.multiprocessing.spawn(run_segmenter_worker, nprocs=nprocs, args=(config_file, kwargs))
```
Comment (mingxin-zheng) [inline auto3dseg/algorithm_templates/segresnet/scripts/segmenter.py:1383]: Hi @Nic-Ma , I remembered you talked about some overhead issue of using `spawn` in some special cases. Can you confirm it won't affect here? Thanks!

Comment (myron) [reply auto3dseg/algorithm_templates/segresnet/scripts/segmenter.py:1383]: the code supports both torchrun .. and direct distributed spawning.  The lines you highlighted will spawn manually (if torchrun wasn't used). But if a user uses torchrun ..., those lines won't be called.  torchrun also uses spawn(). so we actually always use spawn() in distributed training.

**Review by mingxin-zheng — APPROVED** [review]

File: auto3dseg/algorithm_templates/segresnet/scripts/algo.py (hunk 2 above)
Comment (mingxin-zheng) [reply auto3dseg/algorithm_templates/segresnet/scripts/algo.py:248]: @myron Thanks for providing an example on Slack. This is resolved.

File: auto3dseg/algorithm_templates/segresnet/scripts/segmenter.py [hunk 5]
```diff
@@ -0,0 +1,1395 @@
...
+    ResampleToMatchd,
+    SaveImaged,
+    ScaleIntensityRanged,
+    Spacingd,
+    SpatialPadd,
+)
+from monai.utils import MetricReduction, convert_to_dst_type, optional_import, set_determinism
+
+
+class DiceAccuracy:
```
[the remark this answers is not in the document]
Comment (myron) [reply auto3dseg/algorithm_templates/segresnet/scripts/segmenter.py:74]: It's a simple class, that computes dice (compute_dice) and does reduction. The current monai version is over-complicated, it accumulates values, and there is no way to compute just a value of dice easily https://github.com/Project-MONAI/MONAI/issues/5378

File: auto3dseg/algorithm_templates/segresnet/scripts/segmenter.py [hunk 6]
```diff
@@ -0,0 +1,1395 @@
...
+        ts.extend(self.get_custom("after_augment_transforms"))
+
+        return ts
+
+    @classmethod
+    def get_postprocess_transform(
+        cls, save_mask=False, invert=False, transform=None, sigmoid=False, output_path=None
+    ) -> Compose:
+
+        # print('post_process callback:', idx, logits.shape, logits.device)
```
[the remark this answers is not in the document]
Comment (myron) [reply auto3dseg/algorithm_templates/segresnet/scripts/segmenter.py:338]: ok

File: auto3dseg/algorithm_templates/segresnet/scripts/segmenter.py [hunk 7]
```diff
@@ -0,0 +1,1395 @@
...
+        epoch,
+        rank,
+        num_epochs=0,
+        sigmoid=False,
+        use_amp=True,
+        use_cuda=True,
+    ):
+
+        model.train()
+        device = torch.device(rank) if use_cuda else torch.device("cpu")
```
[the remark this answers is not in the document]
Comment (myron) [reply auto3dseg/algorithm_templates/segresnet/scripts/segmenter.py:1078]: no, but it doesn't hurt, and it's convenient for quick debugging on a non-gpu machine.

File: auto3dseg/algorithm_templates/segresnet/configs/hyper_parameters.yaml [hunk 8]
```diff
@@ -1,63 +1,78 @@
...
-# validation
 validate:
-  ckpt_name: "$@bundle_root + '/model_fold' + str(@fold) + '/best_metric_model.pt'"
-  save_mask: true
-  ouptut_path: "$@bundle_root + '/prediction_fold' + str(@fold)"
+  enabled: false
+  ckpt_name: "$@bundle_root + '/model/model.pt'"
+  output_path: "$@bundle_root + '/prediction_validation'"
+  save_mask: false
+  invert: false
```
[the remark this answers is not in the document]
Comment (myron) [reply auto3dseg/algorithm_templates/segresnet/configs/hyper_parameters.yaml:72]: ok

File: auto3dseg/algorithm_templates/segresnet/scripts/segmenter.py [hunk 9]
```diff
@@ -0,0 +1,1395 @@
...
+                raise ValueError("resample_resolution is not provided")
+            pixdim = self.resample_resolution
+            ts.append(
+                Spacingd(
+                    keys=keys,
+                    pixdim=pixdim,
+                    mode=mode,
+                    dtype=torch.float,
+                    min_pixdim=np.array(pixdim) * 0.75,
+                    max_pixdim=np.array(pixdim) * 1.25,
```
[the remark this answers is not in the document]
Comment (myron) [reply auto3dseg/algorithm_templates/segresnet/scripts/segmenter.py:205]: this is to avoid resampling when resolution is similar already.

File: auto3dseg/algorithm_templates/segresnet/scripts/segmenter.py [hunk 10]
```diff
@@ -0,0 +1,1395 @@
...
+                ts.append(
+                    ScaleIntensityRanged(
+                        keys=key, a_min=intensity_bounds[0], a_max=intensity_bounds[1], b_min=-1, b_max=1, clip=False
+                    )
+                )
+                ts.append(Lambdad(keys=key, func=lambda x: torch.sigmoid(x)))
+            elif normalize_mode in ["meanstd", "mri"]:
+                ts.append(NormalizeIntensityd(keys=key, nonzero=True, channel_wise=True))
+            elif normalize_mode in ["pet"]:
+                ts.append(Lambdad(keys=key, func=lambda x: torch.sigmoid((x - x.min()) / x.std())))
```
[the remark this answers is not in the document]
Comment (myron) [reply auto3dseg/algorithm_templates/segresnet/scripts/segmenter.py:244]: this is a heuristic normalization for PET image

File: auto3dseg/algorithm_templates/segresnet/scripts/segmenter.py [hunk 11]
```diff
@@ -0,0 +1,1395 @@
...
+
+    return best_metric
+
+
+def run_segmenter(config_file: Optional[Union[str, Sequence[str]]] = None, **kwargs):
+    """
+    if multiple gpu available, start multiprocessing for all gpus
+    """
+    nprocs = torch.cuda.device_count()
+    if nprocs > 1 and not dist.is_torchelastic_launched():
```
[the remark this answers is not in the document]
Comment (myron) [reply auto3dseg/algorithm_templates/segresnet/scripts/segmenter.py:1366]: I've updated the code to support multi-node

File: auto3dseg/algorithm_templates/segresnet/scripts/segmenter.py [hunk 12]
```diff
@@ -0,0 +1,1395 @@
...
+        if testing_files is None:
+            testing_files, _ = datafold_read(
+                datalist=self.config["data_list_file_path"],
+                basedir=self.config["data_file_base_dir"],
+                fold=-1,
+                key=testing_key,
+            )  # fixme
+        print("testing_files files", len(testing_files))
+
+        inf_loader = self.get_val_loader(data=testing_files, resample_label=False)
```
[the remark this answers is not in the document]
Comment (myron) [reply auto3dseg/algorithm_templates/segresnet/scripts/segmenter.py:1024]: I think it's just a naming misunderstanding, that function get all the validation loading pipeline required for val loader

File: auto3dseg/algorithm_templates/segresnet/scripts/segmenter.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (dongyang0122) [reply auto3dseg/algorithm_templates/segresnet/scripts/segmenter.py:205]: let's say there is one 5mm data point mixed in the 1mm datasets. Will the 5mm data points be resampled? And the descriptions of min_pixdim and max_pixdim is really confusing.

File: auto3dseg/algorithm_templates/segresnet/scripts/segmenter.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (dongyang0122) [reply auto3dseg/algorithm_templates/segresnet/scripts/segmenter.py:244]: do you have the source? thanks!

File: auto3dseg/algorithm_templates/segresnet/scripts/segmenter.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (dongyang0122) [reply auto3dseg/algorithm_templates/segresnet/scripts/segmenter.py:74]: I am fine with it. @wyli @Nic-Ma would you guys like to see the related PR 5378 in this thread?

File: auto3dseg/algorithm_templates/segresnet/scripts/segmenter.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (myron) [reply auto3dseg/algorithm_templates/segresnet/scripts/segmenter.py:244]: it's my own heuristic. thank you

File: auto3dseg/algorithm_templates/segresnet/scripts/segmenter.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (myron) [reply auto3dseg/algorithm_templates/segresnet/scripts/segmenter.py:205]: so if we intend to resample to 1x1x1mm, then the rule above will resample everything outside of the range 0.75x0.75x0.75 - 1.25x1.25x1.25 .   so yes, 5x5x5mm or 1x1x5mm will be resampled to 1x1x1

File: auto3dseg/algorithm_templates/segresnet/scripts/segmenter.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (myron) [reply auto3dseg/algorithm_templates/segresnet/scripts/segmenter.py:1024]: so at the end the function does return the validation loader.    but to get it it also need to initialize data list, transform, etc.  I think this internal function name is ok, or would you like a different name of the function?

[the remark this answers is not in the document]
Comment (dongyang0122) [reply auto3dseg/algorithm_templates/segresnet/scripts/segmenter.py:1024]: maybe create another method `get_infer_loader` or something else

[the remark this answers is not in the document]
Comment (myron) [reply auto3dseg/algorithm_templates/segresnet/scripts/segmenter.py:1024]: there is no difference between validation and inference loading functions in this class.  if I rename it, it won't be appropriate name for validation pipeline, I prefer this name. thank you

File: auto3dseg/algorithm_templates/segresnet/scripts/segmenter.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (wyli) [reply auto3dseg/algorithm_templates/segresnet/scripts/segmenter.py:74]: we'll get questions from the users just like this one https://github.com/Project-MONAI/research-contributions/issues/156
